pkg/scheduler/fair_sharing_iterator.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/util/priority"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// fairSharingIterator orders candidates in a "fair" manner for
+// consideration by scheduling when FairSharing is enabled. See
+// runTournament for description of algorithm.
+type fairSharingIterator struct {
+	// cqToEntry tracks ClusterQueues which still have workloads
+	// to schedule, and the corresponding workload entry.
+	cqToEntry     map[*cache.ClusterQueueSnapshot]*entry
+	entryComparer entryComparer
+}
+
+func makeFairSharingIterator(entries []entry, workloadOrdering workload.Ordering) *fairSharingIterator {
+	f := fairSharingIterator{
+		cqToEntry: make(map[*cache.ClusterQueueSnapshot]*entry, len(entries)),
+		entryComparer: entryComparer{
+			drsValues:        make(map[drsKey]int),
+			workloadOrdering: workloadOrdering,
+		},
+	}
+	for i := range entries {
+		f.cqToEntry[entries[i].clusterQueueSnapshot] = &entries[i]
+	}
+	return &f
+}
+
+func (f *fairSharingIterator) hasNext() bool {
+	return len(f.cqToEntry) > 0
+}
+
+func (f *fairSharingIterator) pop() *entry {
+	cq := f.getCq()
+
+	// CQ has no Cohort. We simply return its workload.
+	if !cq.HasParent() {
+		entry := f.cqToEntry[cq]
+		delete(f.cqToEntry, cq)
+		return entry
+	}
+
+	// CQ is part of a Cohort. We run a tournament, to select the
+	// most fair workload at each level.
+	root := cq.Parent().Root()
+	f.entryComparer.computeDRS(root, f.cqToEntry)
+	entry := runTournament(root, f.entryComparer, f.cqToEntry)
+
+	delete(f.cqToEntry, entry.clusterQueueSnapshot)
+	return entry
+}
+
+// getCq returns a CQ with a workload pending scheduling.
+func (f *fairSharingIterator) getCq() *cache.ClusterQueueSnapshot {
+	for cq := range f.cqToEntry {
```

Comment:
This picks up the "first" element from hashmap, but I believe this may depend on the PRNG: https://pkg.go.dev/hash/maphash. 

This is probably not a problem for the algorithm e2e, because IIUC the order of processing the root cohorts would not matter, as they use disjoint resource pools. 

Mostly I want to confirm my understanding, but if this is the case I would suggest to either add a comment why this is safe e2e or make sure a deterministic element is chosen.

Reply:
I'll add a comment. While the scheduling order is non-deterministic, scheduling is (nearly*) deterministic within any given Cohort. I don't think we should rely on scheduling order being deterministic outside of the Cohort scope

*edge case where DRF, priority, and timestamp are the same. We don't have a UUID tiebreak as we have in preemption

Reply:
sgtm